### PR TITLE
Make orml oracle instantiable

### PR DESCRIPTION
**Pull request by prin-r**
[see specification](https://github.com/open-web3-stack/open-runtime-module-library/issues/245)

Orml

- ✅ Make orml-oracle instantiable
- ✅ trait MultiDataProvider<ProviderId, Key, Value>
  - fn get(source: ProviderId, key: &Key) -> Option<Value>
- ✅ macro! create_midean_value_data_provider(TypeName:ident, $(Providers:ty),+)
  - takes list of type that impl DataProvider
  - create a type struct $TypeName<MinimumCount: Get<u32>, ExpiresIn: Get<Moment>, Key, Value: Ord>
  - impl DataProvider like the DefaultCombineData but read values from $Providers

### CODE REVIEW

File: traits/src/lib.rs [hunk 1]
```diff
@@ -56,3 +64,127 @@ pub enum Change<Value> {
...
+							timestamp: (x.timestamp + y.timestamp) / 2u64,
+						}),
+						_ => None,
+					}
+				} else {
+					Some(values[mid])
+				}
+			}
+
+			fn get_all_values() -> Vec<(CurrencyId, Option<TimestampedValue<Price, Moment>>)> {
```
Comment (shaopengw) [inline traits/src/lib.rs:140]: Not 100% sure, should we also find median in `get_all_values`? @xlc

**Comment by shaopengw** [comment]
Also needs unit tests to cover changes in this PR.

File: traits/src/lib.rs [hunk 2]
```diff
@@ -56,3 +64,127 @@ pub enum Change<Value> {
...
+				)*
+
+				values.retain(|&x| x != None);
+				values.sort_by(|a, b| a.cmp(&b));
+				if values.len() == 0 {
+					return None;
+				}
+				let mid = values.len() / 2;
+				if values.len() % 2 == 0 {
+					match (values[mid - 1], values[mid]) {
```
[the remark this answers is not in the document]
Comment (prin-r) [reply traits/src/lib.rs]: @shaopengw , I have added a comment as your suggested.
Now, the rest is to add units test.

**Comment by prin-r** [comment]
@xlc , @shaopengw 

As discussed on discord, I'm handing this off to you guys. Please take this forward by adding tests. Let me know if I need to do anything.

File: traits/src/data_provider.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+/// Extended data provider to provide timestamped data by key with no-op, and
+/// all data.
+pub trait DataProviderExtended<Key, TimestampedValue> {
+	/// Get timestamped value by key
+	fn get_no_op(key: &Key) -> Option<TimestampedValue>;
+	/// Provide a list of tuples of key and timestamped value
+	fn get_all_values() -> Vec<(Key, Option<TimestampedValue>)>;
+}
+
+#[allow(dead_code)] // rust cannot defect usage in macro_rules
```
Comment (xlc) [inline traits/src/data_provider.rs:25]:
```suggestion
#[allow(dead_code)] // rust cannot detect usage in macro_rules
```

File: traits/src/data_provider.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	Some(median.clone())
+}
+
+#[macro_export]
+macro_rules! create_median_value_data_provider {
+	($name:ident, $key:ty, $value:ty, $timestamped_value:ty, [$( $provider:ty ),*]) => {
+		pub struct $name;
+		impl $crate::DataProvider<$key, $value> for $name {
+			fn get(key: &$key) -> Option<$value> {
+				let mut values = vec![];
```
Comment (xlc) [inline traits/src/data_provider.rs:43]:
```suggestion
				let values = vec![ $(<$provider as $crate::DataProvider<$key, $value>>::get(&key)),* ];
```

File: traits/src/data_provider.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+					}
+				)*
+				$crate::data_provider::median(values)
+			}
+			fn get_all_values() -> Vec<($key, Option<$timestamped_value>)> {
+				let mut keys = sp_std::collections::btree_set::BTreeSet::new();
+				$(
+					<$provider as $crate::DataProviderExtended<$key, $timestamped_value>>::get_all_values()
+						.into_iter()
+						.for_each(|(k, _)| { keys.insert(k); });
```
Comment (xlc) [inline traits/src/data_provider.rs:67]: not sure if `.map(|(k, _) k|).collect::<BTreeSet>::()` will work

**Review by xlc — APPROVED** [review]

File: traits/src/data_provider.rs (hunk 4 above)
Comment (shaopengw) [reply traits/src/data_provider.rs:43]: The `if let` below are to filter out `None` values, or we have to do a `retain`.

File: traits/src/data_provider.rs (hunk 5 above)
Comment (shaopengw) [reply traits/src/data_provider.rs:67]: Hmm, I tried, it works but doesn't read as straightforward as the previous way IMO, and also had to introduce an extra vec:

```rust
let keys = vec![$( <$provider as $crate::DataProviderExtended<$key, $timestamped_value>>::get_all_values() ),*]
    .into_iter()
    .flatten()
    .map(|(k, _)| k).collect::<sp_std::collections::btree_set::BTreeSet<_>>();
```

**Pull request merged**